Mathlib/Combinatorics/Additive/SubsetSum.lean

```diff
@@ -0,0 +1,117 @@
+/-
+Copyright (c) 2025 Aviv Bar Natan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Aviv Bar Natan
+-/
+import Mathlib.Algebra.BigOperators.Group.Finset.Basic
+import Mathlib.Algebra.Order.Monoid.Canonical.Defs
+import Mathlib.Data.Finset.Basic
+import Mathlib.Data.Finset.Card
+import Mathlib.Data.Finset.Defs
+import Mathlib.Data.Finset.Image
+import Mathlib.Data.Finset.Powerset
+import Mathlib.Data.Finset.Max
+import Mathlib.Data.Nat.Choose.Basic
+
+/-!
+# Subset sums
+
+This file defines the subset sum of a finite subset of a commutative monoid.
+
+## Main results
+
+* `SubsetSum.zero_mem`: Zero is always in the subset sum.
+* `SubsetSum.set_subset_SubsetSum`: The original set is a subset of its subset sum.
+* `SubsetSum.monotone`: Subset sum is monotone with respect to set inclusion.
+* `SubsetSum.card_lower_bound`: A lower bound on the cardinality of the subset sum
+  in a canonically linearly ordered commutative monoid with cancellation.
+
+## References
+
+* [Melvyn B. Nathanson, *Inverse theorems for subset sums*][Nathanson1995]
+-/
+
+variable {α : Type*} [DecidableEq α] [AddCommMonoid α]
+
+/-- The SubsetSum of a finite set `A` in a commutative monoid is the set of all sums
+  of subsets of `A`. -/
+def SubsetSum (A : Finset α) : Finset α := A.powerset.image fun S ↦ S.sum id
+
+namespace SubsetSum
+
```

Comment:
The philosophy of "API" says that whenever you make a new definition, you should *immediately* make the characteristic predicate to interact with the definition. In this case, the API is
```lean
theorem mem_subsetSum_iff {x : α} : x ∈ s.subsetSum ↔ ∃ a ⊆ s, a.sum id = x := by simp [subsetSum]
```

Reply:
You can see how much it saves you below from writing `mem_image` and `mem_powerset` 1000 times.